[iOS & tvOS] Trailers

JPKribs opened the pull request:
### Summary

Resolves: https://github.com/jellyfin/Swiftfin/issues/1108

This PR introduces trailer support for both iOS and tvOS clients along with several UI improvements to enhance user experience across both platforms.

---

### Trailer Support
- Added trailer functionality for both iOS and tvOS
- Dynamic trailer button that displays all available trailers
- Automatically hides when no trailers are available
- Single trailers play immediately; multiple trailers display in a menu
- External trailers show a warning before leaving the app
- Local trailers play within the user's selected player
- Added user preference options to filter trailers (all, none, external only, local only)

### UI Improvements
- Created a unified `ActionButton` struct, mirroring the existing tvOS implementation
- Migrated all existing buttons to this new `ActionButton` struct for consistency

### tvOS-Specific Enhancements
- Redesigned action button layout to accommodate the trailer button
- Moved version selector to the same row as play button
- Play button now dynamically resizes to fit the version button when multiple versions exist
- Organized action buttons (play, mark as played, favorite, trailers, advanced options) for better spacing
- Expanded play button width from 400px to 440px to ensure 100x100 is possible for all full action buttons
- Added YouTube app integration for external trailers with appropriate error handling
- Updated tvOS play button titles to show the currently selected version when multiple versions exist
- Scroll the selected version on tvOS automatically if it exceeds the area allocated to it

---

### Settings Screenshot

<details>

<summary>iOS & iPadOS</summary>

<img width="323" alt="iOS - Settings" src="https://github.com/user-attachments/assets/77ccbd0a-8158-41dc-86d4-3d1598208474" />

</details>

<details>

<summary>tvOS</summary>

![Simulator Screenshot - Apple TV 4K (3rd generation) - 2025-03-20 at 19 40 52](https://github.com/user-attachments/assets/adff94b9-6c7e-4915-8f9b-70c65c1b2580)

</details>

---

### iOS Videos

<details>

<summary>iOS & iPadOS - One Trailer Available</summary>

https://github.com/user-attachments/assets/c494d9a6-abf1-4ab4-b4af-a79d02009341

</details>

<details>

<summary>iOS & iPadOS - Multiple Trailers Available</summary>

https://github.com/user-attachments/assets/55a03158-0e8b-4e70-8bde-fc6cfc95c121

</details>

<details>

<summary>tvOS</summary>

https://github.com/user-attachments/assets/f7c48770-f6d9-4eba-bb1a-4b0e39555813

</details>

---

### tvOS Button Options

<details>

<summary>tvOS - 2 Buttons</summary>

![tvOS - 2 Buttons](https://github.com/user-attachments/assets/bf482b8a-b960-41c8-a9bb-24e1d9c0e53d)

</details>

<details>

<summary>tvOS - 2.5 Buttons</summary>

![tvOS - 2 5 Buttons](https://github.com/user-attachments/assets/d328bc94-5468-4af0-815b-a949f7954bc6)

</details>

<details>

<summary>tvOS - 3 Buttons</summary>

![tvOS - 3 Buttons](https://github.com/user-attachments/assets/a211e635-af3d-44bc-9ebe-590f1a6e7f65)

</details>

<details>

<summary>tvOS - 3.5 Buttons</summary>

![tvOS - 3 5 Buttons](https://github.com/user-attachments/assets/c7a1d9ea-9b31-43a8-a864-87a603adf6a7)

</details>

---

### tvOS Videos

<details>

<summary>tvOS - Version Name Handling</summary>

https://github.com/user-attachments/assets/c7bf4799-83a1-4aff-81d6-b2ae9cd8519e

</details>

<details>

<summary>tvOS - One Trailer Available</summary>

https://github.com/user-attachments/assets/e2a07eb3-7a2f-4580-a58e-593d19f05da2

</details>

<details>

<summary>tvOS - Multiple Trailers Available</summary>

https://github.com/user-attachments/assets/7e3c29a2-214f-4c5d-befb-4311adbdcb46

</details>

JPKribs commented:
@LePips I pretty much have this done for iOS/iPadOS. Something I am realizing... Can we play remote trailers on tvOS? Like, a YouTube Video? It looks like I can do an OpenURL for one but I don't know if/how it works from the emulator. Would it open the Youtube App if it's installed? On iOS, worst case, it opens Safari if the app isn't installed. tvOS it just fails.

How do you think we should handle that? Just print an error to the screen or should we hide remote trailers on tvOS?

JPKribs commented:
So... Trailers opens us up to a point where there might be more than 4 items in a row. I took this as an opportunity to try and build a layout. I eventually just let ChatGPT take a shot at it and it works! That being said, I want to clean it up but I wanted to verify this was what we would want for this layout. This isn't a requirement for the Trailers PR since that is done in >200 new lines but Trailers + Versions push us outside the bounds of the screen.

Below are examples of various combinations:

![Simulator Screenshot - Apple TV 4K (3rd generation) - 2025-03-18 at 01 13 38](https://github.com/user-attachments/assets/d10ee7a0-fcf2-47d7-9473-52b2f8812f30)

![Simulator Screenshot - Apple TV 4K (3rd generation) - 2025-03-18 at 01 45 34](https://github.com/user-attachments/assets/49f2366e-60d1-4e41-8de2-f95dff56c734)

![Simulator Screenshot - Apple TV 4K (3rd generation) - 2025-03-18 at 01 46 07](https://github.com/user-attachments/assets/03e7b303-51da-471b-a2ed-a35136dbe11f)

![Simulator Screenshot - Apple TV 4K (3rd generation) - 2025-03-18 at 01 46 59](https://github.com/user-attachments/assets/e84ed750-64fa-4367-8053-d244444d851c)

JPKribs commented:
@LePips Sorry about this! I see you're doing some cleanup and I don't want my commit to throw you off. I added some logic to combine the `ActionButton` and `ActionMenu` to make it easier to style them at the same time if anything changes. I don't know if you'd prefer that so I'll just include my changes below to avoid causing any commit conflicts on your end:

<details>

<summary>This replaces `ActionButton`. You can delete `ActionMenu`.</summary>

```
//
// Swiftfin is subject to the terms of the Mozilla Public
// License, v2.0. If a copy of the MPL was not distributed with this
// file, you can obtain one at https://mozilla.org/MPL/2.0/.
//
// Copyright (c) 2025 Jellyfin & Jellyfin Contributors
//

import SwiftUI

extension ItemView {

    struct ActionButton<Content: View>: View {

        // MARK: - Mode

        private enum Mode {
            case button
            case menu
        }

        // MARK: - Environment Objects

        @Environment(\.isSelected)
        private var isSelected

        // MARK: - Focus State

        @FocusState
        private var isFocused: Bool

        // MARK: - Required Properties

        private let mode: Mode
        private let title: String
        private let icon: String

        // MARK: - Button Properties

        private let selectedIcon: String?
        private let onSelect: (() -> Void)?

        // MARK: - Menu Properties

        private let content: (() -> Content)?

        // MARK: - Body

        var body: some View {
            Group {
                switch mode {
                case .button:
                    Button(action: onSelect!) {
                        labelView
                    }
                case .menu:
                    Menu {
                        content?()
                    } label: {
                        labelView
                    }
                }
            }
            .focused($isFocused)
            .buttonStyle(ActionButtonStyle())
        }

        // MARK: - Label Views

        private var labelView: some View {
            ZStack {
                if mode == .button && isSelected {
                    RoundedRectangle(cornerRadius: 10)
                        .fill(
                            isFocused ? AnyShapeStyle(HierarchicalShapeStyle.primary) :
                                AnyShapeStyle(HierarchicalShapeStyle.primary.opacity(0.5))
                        )
                } else {
                    RoundedRectangle(cornerRadius: 10)
                        .fill(isFocused ? Color.white : Color.white.opacity(0.5))
                }

                Label(title, systemImage: (mode == .button && isSelected) ? (selectedIcon ?? icon) : icon)
                    .hoverEffectDisabled()
                    .focusEffectDisabled()
                    .font(.title3)
                    .fontWeight(.semibold)
                    .foregroundStyle(.black)
                    .labelStyle(.iconOnly)
            }
            .accessibilityLabel(title)
        }
    }
}

// MARK: - Initializers

extension ItemView.ActionButton {

    // MARK: Button Initializer

    init(
        _ title: String,
        icon: String,
        selectedIcon: String,
        onSelect: @escaping () -> Void
    ) where Content == EmptyView {
        self.mode = .button
        self.title = title
        self.icon = icon
        self.selectedIcon = selectedIcon
        self.onSelect = onSelect
        self.content = nil
    }

    // MARK: Menu Initializer

    init(
        _ title: String,
        icon: String,
        @ViewBuilder content: @escaping () -> Content
    ) {
        self.mode = .menu
        self.title = title
        self.icon = icon
        self.selectedIcon = nil
        self.onSelect = nil
        self.content = content
    }
}

struct ActionButtonStyle: ButtonStyle {

    // MARK: - Environment Properties

    @Environment(\.isFocused)
    private var isFocused

    // MARK: - Body

    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .scaleEffect(scaleEffect(isPressed: configuration.isPressed))
            .animation(
                .easeInOut(
                    duration: configuration.isPressed ? 0.2 : 0.15
                ),
                value: configuration.isPressed || isFocused
            )
    }

    // MARK: - Scale Effect Size

    private func scaleEffect(isPressed: Bool) -> CGFloat {
        if isPressed {
            return 0.9
        } else if isFocused {
            return 1.1
        } else {
            return 1.0
        }
    }
}
```

</details>

JPKribs commented:
Okay! I'm finally done screwing with the button style. The `Layout` works *(if we like how it looks)*, the buttons are all styled in the same exact way so there should be no different between any of them, and the focus/press scaling is a flat 16/4 so it should be consistent whether the button is 100px or 400px. Below is what this looks like.

There is a single YouTube Trailer for this. This also shows my error handling for trying to open Youtube when it's not present:

<details>

<summary>2 Buttons</summary>

https://github.com/user-attachments/assets/ae2a958c-8a87-444e-bc92-f915a26bb9e1

</details>

<details>

<summary>3 Buttons</summary>

https://github.com/user-attachments/assets/807d9823-9a5f-4980-ba20-8db1ee64e4e7

</details>

<details>

<summary>4 Buttons</summary>

https://github.com/user-attachments/assets/649e6fa1-987b-4dab-a48f-24740a783439

</details>

<details>

<summary>5 Buttons</summary>

https://github.com/user-attachments/assets/9dfd4eca-e10d-4b28-9479-882a505c102e

</details>

LePips reviewed (changes requested):
We have to stick to 2 rows in the tvOS play button + action section. I do really like the design where the menu is a smaller view on the trailing side. With this, it looks like it's time to look at implementing the version selection as a trailing menu next to the play button.

https://github.com/jellyfin/Swiftfin/pull/1429#pullrequestreview-2627722684

I'm a bit iffy on external trailers but that's just me. An option to not present external trailers sounds good to me.

We can go a step further on the trailers such that if there is only one valid trailer to present we instead launch that trailer over making them select through the menu.

JPKribs commented:
I'm going to work on this more tomorrow but this is my latest version:

<img width="326" alt="Screenshot 2025-03-19 at 22 39 29" src="https://github.com/user-attachments/assets/aabf7227-2b53-4764-8653-c46298efbd77" />

Changes:

1. Moved the Version Selection (If needed) to the PlayButton
2. Make the Trailers button select the only trailer if there was a single trailer. Menu still appears if there are 2+ valid trailers
3. Hide remote trailed if disabled. Still need to add a setting for that one in both iOS and tvOS.

JPKribs commented:
I don't know if I over engineered this... I've updated the `PlayButton` to use the Version title if there are multiple versions. This should scroll the text if it's larger than it's container:

https://github.com/user-attachments/assets/03dd6829-dcf3-4da9-be8b-69781c045f77

I also updated the PlayButton and ActionHStack to be 440 instead of 400. The reason for this is because the advanced options is 60 and spacing is 20

100+20 per button. Since it's 100 high 100 wide makes a square at max population. 120 * 3 == 360 + 60 = 420 + 20 for the spacing of that last item. 

Then, for the play button and version button it's 440 or 320 if the version exists.

LePips commented on Swiftfin tvOS/Views/ItemView/Components/ActionButton/Components/ActionButtonScaleModifier.swift:
```diff
@@ -0,0 +1,50 @@
+
+import SwiftUI
+
+struct ActionButtonScaleModifier: ViewModifier {
```
This doesn't scale down when the button is selected.

JPKribs replied:
Ugh, I agree with what you previously said about not wanting to do custom `ButtonStyles` on tvOS. I have this working now but please let me know if this is completely over-engineered / not worth doing. The issue is using the `isPressed` modifier. `.card` buttons still finish their animation even if the button is just tapped. I don't love this but it seems to work:

https://github.com/user-attachments/assets/7af4596c-654b-448a-95fd-3ff56251f738

---

I'm open to a better solution! The alternative to all this is to use `.card` for he buttons then try to mirror `.card` as close as possible for menus. I wasn't having a lot of luck with that last time but I'd be fine to try again if we don't like this:

```
struct ActionButtonScaleModifier: ViewModifier {

    // MARK: - Button Properties

    let size: CGSize

    // MARK: - Expansion Properties

    let expansion: CGFloat
    let animationDuration: Double

    // MARK: - Expansion Reason(s)

    let isFocused: Bool
    let isPressed: Bool

    // MARK: - Internal State

    @State
    private var currentScale: CGFloat = 0.0
    @State
    private var isAnimating: Bool = false
    @State
    private var animationTask: Task<Void, Never>? = nil

    // MARK: - Body

    func body(content: Content) -> some View {
        let baseSize = max(size.width, size.height)

        return content
            .scaleEffect((baseSize + (2 * currentScale)) / baseSize)
            .onChange(of: isPressed) { _, newValue in
                animationTask?.cancel()

                if newValue {
                    withAnimation(.easeInOut(duration: animationDuration / 2)) {
                        currentScale = expansion
                    }

                    animationTask = Task {
                        try? await Task.sleep(nanoseconds: UInt64(animationDuration * 0.5 * 1_000_000_000))

                        if !Task.isCancelled {
                            await MainActor.run {
                                if !isPressed {
                                    withAnimation(.easeInOut(duration: animationDuration)) {
                                        currentScale = isFocused ? expansion : 0.0
                                    }
                                }
                            }
                        }
                    }
                } else {
                    animationTask = Task {
                        await MainActor.run {
                            withAnimation(.easeInOut(duration: animationDuration)) {
                                currentScale = isFocused ? expansion : 0.0
                            }
                        }
                    }
                }
            }
            .onChange(of: isFocused) { _, newValue in
                if !isPressed {
                    withAnimation(.easeInOut(duration: animationDuration)) {
                        currentScale = newValue ? expansion : 0.0
                    }
                }
            }
            .onAppear {
                currentScale = isFocused ? expansion : 0.0
            }
    }
}

```

LePips replied:
> The alternative to all this is to use .card for he buttons then try to mirror .card as close as possible for menus.

Let's please do this because then it wouldn't require handling the pressed state.

JPKribs replied:
Sounds good. I've gotten something very close to this. Since Menu don't have `.card` or the same `ButtonStyle` options, I had to try and mirror native similar to https://github.com/jellyfin/Swiftfin/pull/1451. Open to feedback if you can think of anything. This passes the test that most people I've shown don't notice the difference but if you stare at it too long you'll notice the Menu's are just slightly less rounded in certain layouts.

Screenshots and video below:

https://github.com/user-attachments/assets/ec1ecd35-a603-405e-82a9-4f93c76acbab

![Simulator Screenshot - Apple TV 4K (3rd generation) - 2025-04-03 at 14 29 43](https://github.com/user-attachments/assets/62b1126f-6afa-4b28-84a9-cb98a6d9396d)

![Simulator Screenshot - Apple TV 4K (3rd generation) - 2025-04-03 at 14 30 07](https://github.com/user-attachments/assets/346e7ffe-35d1-4fbf-a2c5-d641668add57)

JPKribs replied:
Sorry, last item. Adding `.scrollClipDisabled()` to the `SeriesEpisodeSelector.SeasonsHStack` fixes a shadow issue. 1) any concerns using `.scrollClipDisabled()` there? 2) do you want me to include this here while I'm on this view or open another PR? I realize I have way too many open PRs right now so I don't want to pile more on your plate if I can avoid it:

Main
<img width="323" alt="Screenshot 2025-04-03 at 14 58 14" src="https://github.com/user-attachments/assets/57a9cfd3-3dd4-44d8-9e37-ef14c18fa8fa" />

scrollClipDisabled
<img width="325" alt="Screenshot 2025-04-03 at 14 58 46" src="https://github.com/user-attachments/assets/8656dedd-31b5-4c4e-ba8f-ec248ba03858" />

LePips replied:
Do so in a smaller PR, there are probably a few other areas that could be added.

LePips commented on Translations/en.lproj/Localizable.strings:
```diff
@@ -1012,6 +1018,12 @@
 "learnMoreEllipsis" = "Learn more...";
 
+/// Leaving Swiftfin
+"leavingSwiftfin" = "Leaving Swiftfin";
```
I won't care for strings like we're leaving the app in some secure context.

JPKribs replied:
Remove just the text or the text and the confirmation dialog? Confirmation dialog I was using as a catch because I find it kind of annoying when one app abruptly kicks me out into another app.

I can go either way though since Trailers there is more of an expectation especially since it's `Remote`

LePips replied:
I've removed the dialog, so just the string here.

LePips commented on Translations/en.lproj/Localizable.strings:
```diff
@@ -1012,6 +1018,12 @@
+"leavingSwiftfin" = "Leaving Swiftfin";
+
+/// This will open content outside of Swiftfin. Are you sure you want to continue?
+"leavingSwiftfinWarning" = "This will open content outside of Swiftfin. Are you sure you want to continue?";
```
I won't care for strings like we're leaving the app in some secure context.

JPKribs commented on Swiftfin tvOS/Components/ScrollingText.swift:
```diff
@@ -0,0 +1,262 @@
+
+import SwiftUI
+
+public struct ScrollingText: View {
```
Good call! I got ahead of myself here. Removed in next commit.

LePips reviewed (approved):
Thanks for working through with me on this! I'll get this out to TestFlight shortly.
